include/SFML/Graphics/Shader.hpp

```diff
@@ -457,6 +457,18 @@ class SFML_GRAPHICS_API Shader : GlResource, NonCopyable
     void setParameter(const std::string& name, CurrentTextureType);
 
     ////////////////////////////////////////////////////////////
+    /// \brief Get the underlying OpenGL handle of the shader.
+    ///
+    /// You shouldn't need to use this function, unless you have
+    /// very specific stuff to implement that SFML doesn't support,
+    /// or implement a temporary workaround until a bug is fixed.
+    ///
+    /// \return OpenGL handle of the shader or 0 if not yet loaded
+    ///
+    ////////////////////////////////////////////////////////////
+    unsigned int getNativeHandle() const;
```

Comment:
We can only return a single type in the interface, unless we expose the GLLoader header, which we don't want.

Theoretically the user should be able to cast from `unsigned int` to whatever handle type they are using. The compiler will complain a lot if you aren't careful, but other than that it shouldn't really be a problem. The people who know how to use the handle will be able to take care of this themselves, after all the documentation already discourages people who don't know what they are doing from using this anyway :wink:.

Side note: The casting is really only a problem in legacy contexts on OS X because shaders are an extension. In core contexts (with shaders as part of the core specification), the OpenGL specification specifies the type of the handle for us (which is the same as all the other handles), so Apple isn't able to be creative any more :stuck_out_tongue:.